include/swift/AST/DiagnosticsFrontend.def

```diff
@@ -350,6 +350,15 @@ ERROR(error_opening_explicit_module_file,none,
       "failed to open explicit Swift module: %0", (StringRef))
 ERROR(error_extracting_flags_from_module_interface,none,
       "error extracting flags from module interface", ())
+WARNING(warning_module_shadowing_may_break_module_interface,none,
+        "public %0 %1 shadows module %2, which is permitted by Swift but may "
+        "stop other compiler versions from importing module %3 or its clients; "
+        "please work around this language bug (SR-898) by renaming either the "
+        "%0 %1 or the module %2, or by passing the '-Xfrontend "
+        "-module-interface-preserve-types-as-written' flags to the Swift "
```

Comment:
IMO we should not recommend a frontend flag to people in a diagnostic. On one hand, frontend flags are not "officially supported", on the other hand, pointing it out in a diagnostic like this may make it sound like it is officially supported (and maybe using other frontend flags is also fine?).

At the very least, if we are going to recommend this flag in the diagnostic, we should expand the help text for the flag to point out what issues it does/does not solve, and that it should only be used as an option of last resort.

Moreover, using the flag might not resolve the issue.  https://bugs.swift.org/browse/SR-14153